projects/034 - Williams/template-sql/09NewAllergies.template.sql

```diff
@@ -37,13 +37,13 @@ FROM #PatientsToInclude;
 IF OBJECT_ID('tempdb..#AllergyAll') IS NOT NULL DROP TABLE #AllergyAll;
 SELECT DISTINCT FK_Patient_Link_ID, TRY_CONVERT(DATE, EventDate) AS EventDate, FK_Reference_Coding_ID
 INTO #AllergyAll
-FROM [RLS].[vw_GP_Events]
+FROM SharedCare.GP_Events
 WHERE (SuppliedCode IN (SELECT Code FROM #AllCodes WHERE (Concept = 'allergy' AND [Version] = 1)))
-      AND EventDate >= @StartDate AND EventDate < @EndDate;
+      AND EventDate < @EndDate;
 
 
 -- Create the table of new allergy code=============================================================================================================
 SELECT FK_Patient_Link_ID AS PatientId, MIN(EventDate) AS Date
 FROM #AllergyAll
-WHERE FK_Patient_Link_ID IN (SELECT FK_Patient_Link_ID FROM #PatientsToInclude)
+WHERE YEAR(EventDate) >= 2019 AND FK_Patient_Link_ID IN (SELECT FK_Patient_Link_ID FROM #PatientsToInclude)
```

Comment:
This actually is the same as before. This SQL is asking to first filter to Dates >= 2019, and then to take the MIN. Instead you want to take the MIN first and then filter out those where the MIN is before 2019. So you need to get rid of `WHERE YEAR(EventDate) >= 2019` and replace it with a `HAVING YEAR(MIN(EventDate)) >=2019` or `HAVING MIN(EventDate) >= '2019-01-01'`. The `HAVING` statement goes after the `GROUP BY` statement. `WHERE` clauses are evaluated before the GROUP BY, while `HAVING` clauses are evaluated afterwards.

Reply:
It is useful to know. Thanks Richard!